### PR TITLE
HPCC-22174 Add LDAP support in envgen2

### PR DESCRIPTION
**Pull request by xwang2713**
Escape "=" with "[equal]" on envgen2 command-line parameter.
Refactor "selector" handling.
Fix envgen2 -in-file processing

The one step configuration is to use envgen2 -in-file <input file in xml format>
The two steps are:
envgen2 -env-in <env 0> -env-out <env 1> -add sw:ldap#<name of ldap>@esp=<esp cluster name>
envgen2 -env-in <env 1> -env-out <env 2> -add-node ldap@ip="10.0.0.10;10.0.2.3"

Signed-off-by Xiaoming Wang <xiaoming.wang@lexisnexis.com>

@kenrowland please help review
I did private builds and will ask UK team to test it.

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Private builds on Jenkins
Will ask US team (Tim Miller and Domenico Pirlo) to test

### CODE REVIEW

**Comment by stuartort** [comment]
@kenrowland, can you code review this

File: deployment/configenv/xml_jlibpt/EnvHelper.cpp [hunk 1]
```diff
@@ -316,7 +316,9 @@ const char* EnvHelper::getXMLTagName(const char* name)
    else if (!strcmp(nameLC, "dropzone"))
       return XML_TAG_DROPZONE;
    else if (!strcmp(nameLC, "eclcc"))
-      return "EclCCProcess";
+      return "EclCCServerProcess";
```
Comment (kenrowland) [inline deployment/configenv/xml_jlibpt/EnvHelper.cpp]: There is disconnect about what this process should be named. The XSD uses EclCCServer, the wizard creates it as EclCCServerProcess. I believe the XSL will accept either. Not sure if a change is necessary here, just wanted to bring it up.

File: deployment/configenv/xml_jlibpt/SWBackupNode.cpp [hunk 2]
```diff
@@ -57,6 +52,5 @@ unsigned SWBackupNode::add(IPropertyTree *params)
        }
        compTree->addPropTree(selector, nodeGroup);
    }
-   return rc;
 }
 }
```
Comment (kenrowland) [inline deployment/configenv/xml_jlibpt/SWBackupNode.cpp]: Is a failure here, the property tree not added, anything that needs to be handled?

File: deployment/configenv/xml_jlibpt/SWEspProcess.cpp [hunk 3]
```diff
@@ -26,16 +26,8 @@ SWEspProcess::SWEspProcess(const char* name, EnvHelper * envHelper):SWProcess(na
 {
 }
 
-unsigned SWEspProcess::add(IPropertyTree *params)
+void SWEspProcess::addOtherSelector(IPropertyTree *compTree, IPropertyTree *params)
```
Comment (kenrowland) [inline deployment/configenv/xml_jlibpt/SWEspProcess.cpp]: is a failure in adding the binding something to worry about?

File: deployment/configenv/xml_jlibpt/SWLdapProcess.hpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+{
+
+class SWLdapProcess : public SWProcess
+{
+public:
+    SWLdapProcess(const char* name, EnvHelper * envHelper);
+
+    unsigned add(IPropertyTree *params);
+    //IPropertyTree * addComponent(IPropertyTree *params);
+    //void addBinding(IPropertyTree *parent, IPropertyTree * attrs);
```
Comment (kenrowland) [inline deployment/configenv/xml_jlibpt/SWLdapProcess.hpp]: Should the commented code lines remain?

File: deployment/envgen2/EnvGen.cpp [hunk 5]
```diff
@@ -933,7 +942,7 @@ void CEnvGen::addUpdateTaskFromFile(const char * inFile)
    }
```
Comment (kenrowland) [inline deployment/envgen2/EnvGen.cpp]: Should 
  if (!(*propName)) continue;
Be
  if ( (propName != nullptr) && *propName) continue;

File: deployment/envgen2/EnvGen.cpp [hunk 6]
```diff
@@ -922,7 +931,7 @@ void CEnvGen::addUpdateTaskFromFile(const char * inFile)
...
 
    // add Config attributies to params
-   IPropertyTree *pCfg = inPTree->queryPropTree("Config");
+   IPropertyTree *pCfg = m_params->queryPropTree("Config");
```
Comment (kenrowland) [inline deployment/envgen2/EnvGen.cpp]: Can pCfg be nullptr?

**Review by kenrowland — COMMENTED** [review]
@xwang2713 Please see comments

File: deployment/configenv/xml_jlibpt/EnvHelper.cpp (hunk 1 above)
Comment (xwang2713) [reply deployment/configenv/xml_jlibpt/EnvHelper.cpp]: Let me double check. I thought it EclCCServerProcess so I made the change

File: deployment/configenv/xml_jlibpt/SWBackupNode.cpp (hunk 2 above)
Comment (xwang2713) [reply deployment/configenv/xml_jlibpt/SWBackupNode.cpp]: A exception should be throw in method which return rc

File: deployment/configenv/xml_jlibpt/SWEspProcess.cpp (hunk 3 above)
Comment (xwang2713) [reply deployment/configenv/xml_jlibpt/SWEspProcess.cpp]: Throw exception

File: deployment/configenv/xml_jlibpt/SWLdapProcess.hpp (hunk 4 above)
Comment (xwang2713) [reply deployment/configenv/xml_jlibpt/SWLdapProcess.hpp]: removed

File: deployment/envgen2/EnvGen.cpp (hunk 5 above)
Comment (xwang2713) [reply deployment/envgen2/EnvGen.cpp]: It checks propName string is not empty string. I really reply on PropertyTreeIterator to throw exception if no property name item available

File: deployment/envgen2/EnvGen.cpp (hunk 6 above)
Comment (xwang2713) [reply deployment/envgen2/EnvGen.cpp]: add assert(pCfg);

**Comment by xwang2713** [comment]
@kenrowland  update for review comments. Please help review again. Thanks

**Review by kenrowland — APPROVED** [review]
Looks good

**Comment by xwang2713** [comment]
@richardkchapman this is ready to merge.

**Pull request merged**